Marshal/MarshaledObject.swift

```diff
@@ -14,25 +14,33 @@
 import Foundation
 
 
-// MARK: - Types
-
-public typealias MarshaledObject = [String: AnyObject]
-
-
-// MARK: - Dictionary Extensions
+public protocol MarshaledObject {
+    subscript(key: KeyType) -> Any? { get }
+    func anyForKey(key: KeyType) throws -> Any
+    func valueForKey<A: ValueType>(key: KeyType) throws -> A
+    func valueForKey<A: ValueType>(key: KeyType) throws -> A?
+    func valueForKey<A: ValueType>(key: KeyType) throws -> [A]
+    func valueForKey<A: ValueType>(key: KeyType) throws -> [A]?
+    func valueForKey<A: ValueType>(key: KeyType) throws -> Set<A>
+    func valueForKey<A: ValueType>(key: KeyType) throws -> Set<A>?
+    func valueForKey<A: RawRepresentable where A.RawValue: ValueType>(key: KeyType) throws -> A
+    func valueForKey<A: RawRepresentable where A.RawValue: ValueType>(key: KeyType) throws -> A?
+    func valueForKey<A: RawRepresentable where A.RawValue: ValueType>(key: KeyType) throws -> [A]
+    func valueForKey<A: RawRepresentable where A.RawValue: ValueType>(key: KeyType) throws -> [A]?
+    func valueForKey<A: RawRepresentable where A.RawValue: ValueType>(key: KeyType) throws -> Set<A>
+    func valueForKey<A: RawRepresentable where A.RawValue: ValueType>(key: KeyType) throws -> Set<A>?
+}
 
```

Comment:
I was discussing this with @psychoticidiot, trying to guess which methods a user would be likely to override vs. accept the default implementation in the extension. `subscript` for sure needs to be in the protocol declaration. One could make the case for `anyForKey`. It seems unlikely that a user would need to override `valueForKey`. By removing `valueForKey` from the protocol declaration, we might see a performance improvement because it would use static dispatch for those instead of dynamic dispatch. 

What do you think @jarsen ?